### PR TITLE
Add pass to create logical objectfifo link operations

### PR DESCRIPTION
**Pull request by jtuyls**
Adds a pass to create explicit logical objectfifo link operations, linking input and output copy-like operations on these logical objectfifos. This is useful for conversion to the AIE dialect as that relies on similar explicit link operations.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.h [hunk 1]
```diff
@@ -59,6 +59,10 @@ std::unique_ptr<Pass> createAMDAIEControlCodeLoopUnrollPass();
 /// Pass to create a single AIE workgroup.
 std::unique_ptr<Pass> createAMDAIECreateAIEWorkgroupPass();
 
+/// Pass to create logical objectfifo link operations, explicitly linking input
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.h:62]: Nit: linking inputs and outputs. Same for the summary in Passes.td.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+// CHECK-LABEL: func.func @link_two_inputs
+// CHECK:       %[[DMA0:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       %[[DMA1:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       %[[DMA2:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       amdaie.logicalobjectfifo.link
+// CHECK-DAG:   %[[DMA0]]
+// CHECK-DAG:   %[[DMA1]]
+// CHECK-DAG:   %[[DMA2]]
+func.func @link_two_inputs(%arg0: memref<32x1024xi32>, %arg1: memref<32x64xi32, 1>, %arg2: memref<8x8x4x8xi32, 2>) {
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir:28]: Is there also a case linking a input with two outputs?

**Review by yzhang93 — APPROVED** [review]
LGTM. I'm wondering can you create the link when you generate `amdaie.circular_dma_cpy_nd` instead of using a separate pass?

**Comment by jtuyls** [comment]
> LGTM. I'm wondering can you create the link when you generate `amdaie.circular_dma_cpy_nd` instead of using a separate pass?

This link actually duplicates information at this stage, so I don't think it's a good thing to do. However, it's how the AIE dialect object fifos are set up, so it's useful for the conversion. We can maybe get rid of the links at some point, but that would need larger changes on the AIE dialect side as well. 

Also, the pass is just for testing purposes, the conversion pass to AIE will call this function directly: https://github.com/nod-ai/iree-amd-aie/pull/355/files#diff-937907de644696272a7a135dbcdcb3ce05371a4f3e3972f2e9abfd26b5055bfcR23

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir (hunk 2 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir]: Good point, added it.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.h (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.h]: Thanks for the catch.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+/// Utility to add explicit link operations to avoid having to do this during
+/// conversion to AIEDialect operations. This function only consider L2/MT for
+/// links as L1/L3 don't need this linking through AIE objectfifos. Furthermore,
+/// it assumes that all users of a logical objectfifo reside within the same
+/// block and an error will be emitted if that's not the case.
+LogicalResult createLogicalObjectFifoLink(
+    RewriterBase &rewriter,
+    AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjectFifo) {
+  auto memSpace = logicalObjectFifo.getMemorySpace();
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp:24]: `auto` -> `Attribute`.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  // Visit all CopyOp users of this logical objectfifo and add them to
+  // either the input or output side of this logical objectfifo. While
+  // doing this, keep track of the last user operation for insertion
+  // purposes.
+  SmallVector<OpFoldResult> ins;
+  SmallVector<OpFoldResult> outs;
+  CopyOpInterface lastUserOp;
+  for (auto userOp : logicalObjectFifo->getUsers()) {
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp:36]: `auto` -> `Operation*`.

**Review by Abhishek-Varma — APPROVED** [review]
LGTM!

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp (hunk 3 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp (hunk 4 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp]: Done

**Pull request merged**